winch/codegen/src/trampoline.rs

```diff
@@ -0,0 +1,163 @@
+        let func_ptr_offset = trampoline_arg_offsets[2];
+
+        self.masm.mov(
+            val_ptr,
+            RegImm::reg(self.scratch_1),
+            crate::masm::OperandSize::S64,
+        );
+
+        // How much we need to adjust the stack pointer by to account for the alignment
+        // required by the ISA
+        let delta = calculate_frame_adjustment(
+            self.masm.sp_offset(),
+            self.abi.arg_base_offset() as u32,
+            self.abi.call_stack_align() as u32,
+        );
+
+        // The total amount of stack space we need to reserve for the arguments
+        let total_arg_stack_space = align_to(
+            callee_sig.stack_bytes + delta,
+            self.abi.call_stack_align() as u32,
+        );
+
+        self.masm.reserve_stack(total_arg_stack_space);
+
+        // The max size a value can be when reading from the params memory location
+        let value_size = mem::size_of::<u128>();
+
+        callee_sig.params.iter().enumerate().for_each(|(i, param)| {
+            let value_offset = (i * value_size) as u32;
+
+            match param {
+                ABIArg::Reg { reg, ty } => self.masm.load(
+                    self.masm.address_from_reg(self.scratch_1, value_offset),
+                    *reg,
+                    (*ty).into(),
+                ),
+                ABIArg::Stack { offset, ty } => {
+                    self.masm.load(
+                        self.masm.address_from_reg(self.scratch_1, value_offset),
+                        self.scratch_2,
+                        (*ty).into(),
+                    );
+                    self.masm.store(
+                        RegImm::reg(self.scratch_2),
+                        self.masm.address_at_sp(*offset),
+                        (*ty).into(),
+                    );
+                }
+            }
+        });
+
+        // Move the function pointer from it's stack location into a scratch register
+        self.masm.load(
+            self.masm
+                .address_at_sp(self.masm.sp_offset() - func_ptr_offset),
+            self.scratch_1,
+            OperandSize::S64,
+        );
+
+        // Call the function that was passed into the trampoline
+        self.masm.call(CallKind::Indirect(self.scratch_1));
+
+        self.masm.free_stack(total_arg_stack_space);
+
+        // Move the val ptr back into the scratch register so we can load the return values
+        self.masm.load(
+            self.masm
+                .address_at_sp(self.masm.sp_offset() - val_ptr_offset),
+            self.scratch_1,
+            OperandSize::S64,
+        );
+
+        // Move the return values into the value ptr
+        // Only doing a single return value for now
+        let ABIResult::Reg { reg, ty } = &callee_sig.result;
+        self.masm.store(
+            RegImm::reg(*reg),
+            self.masm.address_from_reg(self.scratch_1, 0),
+            (*ty).unwrap().into(),
+        );
+
```

Comment:
Similarly, it might be good to add a TODO here to state that we are missing restoring any callee-saved registers.